File: apps/mobile/modules/safely-push-content/android/src/main/java/expo/modules/safelypushcontent/SafelyPushContentModule.kt
package expo.modules.safelypushcontent

import android.content.Context
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

internal fun walletNamesFile(context: Context): File =
    File(context.noBackupFilesDir, WALLET_NAMES_FILE_NAME)

// Writes the `target_ref -> wallet name` dictionary that SafelyNotificationsService
// reads before a push is displayed. JS never touches the file or its format.
class SafelyPushContentModule : Module() {
    private val context: Context
        get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

    override fun definition() = ModuleDefinition {
        Name("SafelyPushContent")

        AsyncFunction("setWalletNames") { names: Map<String, String> ->
            WalletNameStore(walletNamesFile(context)).save(names)
        }
    }
}

File: apps/mobile/modules/safely-push-content/ios/SafelyPushContentModule.swift
import ExpoModulesCore

internal final class AppGroupUnavailableException: Exception {
    override var reason: String {
        "App Group container \(PushContentCore.appGroupId) is not available"
    }
}

// Writes the `target_ref -> wallet name` dictionary into the App Group container
// so the Notification Service Extension (same PushContentCore, other process)
// can substitute {{wallet_name}} before a push is shown. JS never touches the
// file or its format.
public class SafelyPushContentModule: Module {
    public func definition() -> ModuleDefinition {
        Name("SafelyPushContent")

        AsyncFunction("setWalletNames") { (names: [String: String]) in
            guard let container = FileManager.default.containerURL(
                forSecurityApplicationGroupIdentifier: PushContentCore.appGroupId
            ) else {
                throw AppGroupUnavailableException()
            }
            try PushContentCore.save(walletNames: names, to: PushContentCore.walletNamesURL(in: container))
        }
    }
}
